File: task-64bc3c/environment/app_data/evaluator_beta.R
#!/usr/bin/env Rscript
#
# Reproducibility Evaluator — Implementation Beta
# Alternative evaluation of agent results against stochastic ground truth.
#
# Authors: Research Team B
#

library(jsonlite)

args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3) {
  cat("Usage: Rscript evaluator_beta.R <ground_truth.json> <agent_reports_dir> <output_dir>\n")
  quit(status = 1)
}

gt_path <- args[1]
reports_dir <- args[2]
output_dir <- args[3]
dir.create(output_dir, showWarnings = FALSE, recursive = TRUE)

gt_data <- fromJSON(gt_path, simplifyVector = FALSE)


compute_intervals <- function(gt_results) {
  first <- gt_results[[1]]
  n <- length(gt_results)
  intervals <- list()

  for (key in names(first)) {
    val <- first[[key]]
    if (!is.numeric(val) || length(val) != 1) next

    values <- sapply(gt_results, function(r) r[[key]])
    mean_val <- mean(values)
    std_val <- sd(values)

    if (is.na(std_val) || std_val == 0) {
      intervals[[key]] <- c(mean_val, mean_val)
      next
    }

    # 95th percentile of t-distribution for 95% interval
    t_val <- qt(0.95, n - 1)
    margin <- t_val * std_val * sqrt(1.0 / n)
    intervals[[key]] <- c(mean_val - margin, mean_val + margin)
  }

  return(intervals)
}


coerce_value <- function(val) {
  if (is.numeric(val)) return(val)
  if (is.character(val)) {
    result <- suppressWarnings(as.numeric(trimws(val)))
    if (!is.na(result)) return(result)
    return(val)
  }
  return(val)
}


evaluate_capsule <- function(gt_capsule, reported) {
  gt_results <- gt_capsule$results
  first <- gt_results[[1]]
  intervals <- compute_intervals(gt_results)

  correct_written <- 0
  correct_vision <- 0
  total_written <- 0
  total_vision <- 0

  for (key in names(first)) {
    gt_val <- first[[key]]

    # Vision question if 'fig' substring appears anywhere in the key
    is_vision <- grepl("fig", key, fixed = TRUE)

    if (is_vision) {
      total_vision <- total_vision + 1
    } else {
      total_written <- total_written + 1
    }

    if (!(key %in% names(reported))) next

    rep_val <- coerce_value(reported[[key]])

    if (is.numeric(gt_val) && length(gt_val) == 1) {
      if (is.numeric(rep_val) && length(rep_val) == 1) {
        bounds <- intervals[[key]]
        if (!is.null(bounds) && rep_val >= bounds[1] && rep_val <= bounds[2]) {
          if (is_vision) {
            correct_vision <- correct_vision + 1
          } else {
            correct_written <- correct_written + 1
          }
        }
      }
    } else if (is.character(gt_val)) {
      # Case-insensitive comparison
      if (tolower(as.character(rep_val)) == tolower(gt_val)) {
        if (is_vision) {
          correct_vision <- correct_vision + 1
        } else {
          correct_written <- correct_written + 1
        }
      }
    } else if (is.list(gt_val) || (is.numeric(gt_val) && length(gt_val) > 1)) {
      if (identical(rep_val, gt_val)) {
        if (is_vision) {
          correct_vision <- correct_vision + 1
        } else {
          correct_written <- correct_written + 1
        }
      }
    }
  }

  return(list(
    capsule_id = gt_capsule$capsule_id,
    correct_written = correct_written,
    correct_vision = correct_vision,
    total_written = total_written,
    total_vision = total_vision
  ))
}


# Evaluate each agent
all_results <- list()
files <- list.files(reports_dir, pattern = "\\.json$", full.names = TRUE)
for (filepath in sort(files)) {
  agent_name <- tools::file_path_sans_ext(basename(filepath))
  report <- fromJSON(filepath, simplifyVector = FALSE)

  capsule_results <- list()
  for (cr in report$capsule_results) {
    gt_capsule <- NULL
    for (g in gt_data) {
      if (g$capsule_id == cr$capsule_id) {
        gt_capsule <- g
        break
      }
    }
    if (is.null(gt_capsule)) next

    reported <- cr$result_report
    if (is.null(reported)) reported <- list()

    result <- evaluate_capsule(gt_capsule, reported)
    capsule_results <- c(capsule_results, list(result))
  }

  all_results[[agent_name]] <- capsule_results
}

# TODO: Implement aggregate scoring per the paper specification
# TODO: Implement prediction interval output file
output <- toJSON(all_results, auto_unbox = TRUE, pretty = TRUE)
writeLines(output, file.path(output_dir, "evaluation_summary.json"))

cat("Beta evaluation complete.\n")

File: task-64bc3c/environment/app_data/evaluator_beta.py
#!/usr/bin/env python3
"""
Reproducibility Evaluator — Implementation Beta
Alternative evaluation of agent results against stochastic ground truth.

Authors: Research Team B
"""

import json
import math
import os
import sys
import numpy as np
from scipy.stats import t


def load_json(filepath):
    with open(filepath) as f:
        return json.load(f)


def compute_intervals(gt_results):
    """Compute 95% intervals for numeric keys across multiple runs."""
    first = gt_results[0]
    n = len(gt_results)
    intervals = {}

    for key, val in first.items():
        if not isinstance(val, (int, float)):
            continue
        values = [r[key] for r in gt_results]
        mean_val = float(np.mean(values))
        std_val = float(np.std(values, ddof=1))

        if std_val == 0.0:
            intervals[key] = (mean_val, mean_val)
            continue

        # 95th percentile of t-distribution for 95% interval
        t_val = t.ppf(0.95, n - 1)
        margin = t_val * std_val * math.sqrt(1.0 / n)
        intervals[key] = (mean_val - margin, mean_val + margin)

    return intervals


def coerce_value(val):
    """Coerce value to numeric type if possible."""
    if isinstance(val, (int, float)):
        return val
    if isinstance(val, str):
        try:
            return float(val.strip())
        except ValueError:
            return val
    return val


def evaluate_capsule(gt_capsule, reported):
    """Evaluate a single capsule result."""
    gt_results = gt_capsule['results']
    first = gt_results[0]
    intervals = compute_intervals(gt_results)

    correct_written = 0
    correct_vision = 0
    total_written = 0
    total_vision = 0

    for key, gt_val in first.items():
        # Vision question if 'fig' substring appears anywhere in the key
        is_vision = 'fig' in key

        if is_vision:
            total_vision += 1
        else:
            total_written += 1

        if key not in reported:
            continue

        rep_val = coerce_value(reported[key])

        if isinstance(gt_val, (int, float)):
            if isinstance(rep_val, (int, float)):
                lower, upper = intervals[key]
                if lower <= rep_val <= upper:
                    if is_vision:
                        correct_vision += 1
                    else:
                        correct_written += 1
        elif isinstance(gt_val, str):
            # Case-insensitive comparison
            if str(rep_val).lower() == gt_val.lower():
                if is_vision:
                    correct_vision += 1
                else:
                    correct_written += 1
        elif isinstance(gt_val, list):
            if rep_val == gt_val:
                if is_vision:
                    correct_vision += 1
                else:
                    correct_written += 1

    return {
        'capsule_id': gt_capsule['capsule_id'],
        'correct_written': correct_written,
        'correct_vision': correct_vision,
        'total_written': total_written,
        'total_vision': total_vision
    }


def main():
    if len(sys.argv) != 4:
        print("Usage: python evaluator_beta.py <ground_truth.json> <agent_reports_dir> <output_dir>")
        sys.exit(1)

    gt_path = sys.argv[1]
    reports_dir = sys.argv[2]
    output_dir = sys.argv[3]
    os.makedirs(output_dir, exist_ok=True)

    gt_data = load_json(gt_path)

    # Evaluate agents
    all_results = {}
    for filename in sorted(os.listdir(reports_dir)):
        if not filename.endswith('.json'):
            continue
        agent_name = filename.replace('.json', '')
        report = load_json(os.path.join(reports_dir, filename))

        capsule_results = []
        for cr in report['capsule_results']:
            gt_capsule = None
            for g in gt_data:
                if g['capsule_id'] == cr['capsule_id']:
                    gt_capsule = g
                    break
            if gt_capsule is None:
                continue
            result = evaluate_capsule(gt_capsule, cr.get('result_report', {}))
            capsule_results.append(result)

        all_results[agent_name] = capsule_results

    # TODO: Implement aggregate scoring per the paper specification
    # TODO: Implement prediction interval output file
    with open(os.path.join(output_dir, 'evaluation_summary.json'), 'w') as f:
        json.dump(all_results, f, indent=2)

    print("Beta evaluation complete.")


if __name__ == '__main__':
    main()
